src/main/scala/xiangshan/frontend/IFU.scala

```diff
@@ -491,8 +509,36 @@ class NewIFU(implicit p: Parameters) extends XSModule
   val f3_af_vec         = RegEnable(f2_af_vec,      f2_fire)
   val f3_pf_vec         = RegEnable(f2_pf_vec ,     f2_fire)
   val f3_gpf_vec        = RegEnable(f2_gpf_vec,     f2_fire)
-  val f3_pc             = RegEnable(f2_pc,          f2_fire)
-  val f3_half_snpc      = RegEnable(f2_half_snpc,   f2_fire)
+
+  val f3_pc_lower_result        = RegEnable(f2_pc_lower_result, f2_fire)
+  val f3_pc_high                = RegEnable(f2_pc_high, f2_fire)
+  val f3_pc_high_plus1          = RegEnable(f2_pc_high_plus1, f2_fire)
+  //val f3_pc             = RegEnable(f2_pc,          f2_fire)
+  val f3_pc             = VecInit(f3_pc_lower_result.map{ i =>
+    Mux(i(pc_adder_cut_point), Cat(f3_pc_high_plus1,i(pc_adder_cut_point-1,0)), Cat(f3_pc_high,i(pc_adder_cut_point-1,0)))})
+
+  val f3_pc_last_lower_result_plus2 = RegEnable(f2_pc_lower_result(PredictWidth - 1) + 2.U, f2_fire)
+  val f3_pc_last_lower_result_plus4 = RegEnable(f2_pc_lower_result(PredictWidth - 1) + 4.U, f2_fire)
+  //val f3_half_snpc      = RegEnable(f2_half_snpc,   f2_fire)
+  
+  /**
+    ***********************************************************************
+    * Half snpc(i) is larger than pc(i) by 4. Using pc to calculate half snpc may be a good choice.
+    ***********************************************************************
+    */
+  val f3_half_snpc      = Wire(Vec(PredictWidth,UInt(VAddrBits.W)))
+  for(i <- 0 until PredictWidth){
+    val j  = f3_pc_last_lower_result_plus2
+    val k  = f3_pc_last_lower_result_plus4
+    if(i == (PredictWidth - 2)){
+      f3_half_snpc(i)   := Mux(j(pc_adder_cut_point), Cat(f3_pc_high_plus1,j(pc_adder_cut_point-1,0)), Cat(f3_pc_high,j(pc_adder_cut_point-1,0)))
+    } else if (i == (PredictWidth - 1)) {
+      f3_half_snpc(i)   := Mux(k(pc_adder_cut_point), Cat(f3_pc_high_plus1,k(pc_adder_cut_point-1,0)), Cat(f3_pc_high,k(pc_adder_cut_point-1,0)))
```

Comment:
3. The naming of `j` and `k` seems a bit unclear here, could be changed along with the previous suggestion(2.).
```scala
  if (i == (PredictWidth - 2)) {
    f3_half_snpc(i) := CatPC(f3_pc_last_lower_result_plus2, f3_pc_high, f3_pc_high_plus1)
  } else if (i == (PredictWidth - 1)) {
    f3_half_snpc(i) := CatPC(f3_pc_last_lower_result_plus4, f3_pc_high, f3_pc_high_plus1)
```